intake_xarray/image.py

```diff
@@ -0,0 +1,265 @@
+    Dask array of all images stacked along the first dimension.  All images
+    will be treated as individual chunks
+    """
+    from dask.array import Array
+    from dask.base import tokenize
+    from dask.bytes.core import OpenFile
+
+    if not imread:
+        from skimage.io import imread
+
+    def _imread(open_file: OpenFile):
+        with open_file as f:
+            return imread(f)
+
+    filenames = [f.path for f in files]
+
+    name = 'imread-%s' % tokenize(filenames, map(get_mtime, filenames))
+
+    with files[0] as f:
+        sample = imread(f)
+    if preprocess:
+        sample = preprocess(sample)
+
+    keys = [(name, i) + (0,) * len(sample.shape)
+            for i in range(len(files))]
+
+    if preprocess:
+        values = [(add_leading_dimension, (preprocess, (_imread, f)))
+                  for f in files]
+    else:
+        values = [(add_leading_dimension, (_imread, f))
+                  for f in files]
+    dsk = dict(zip(keys, values))
+
+    chunks = ((1, ) * len(files), ) + tuple((d, ) for d in sample.shape)
+
+    return Array(dsk, name, chunks, sample.dtype)
+
+
+def reader(file, chunks, imread=None, preprocess=None, **kwargs):
+    """ Read a file object and output an xarray object
+    """
+    import numpy as np
+    from xarray import DataArray
+
+    if not imread:
+        from skimage.io import imread
+
+    with file as f:
+        array = imread(f)
+    if preprocess:
+        array = preprocess(array)
+
+    ny, nx = array.shape[:2]
+    coords = {'y': np.arange(ny),
+              'x': np.arange(nx)}
+    dims = ('y', 'x')
+
+    if len(array.shape) == 3:
+        nband = array.shape[2]
+        coords['band'] = np.arange(nband)
+        dims += ('band',)
+
+    return DataArray(array, coords=coords, dims=dims).chunk(chunks=chunks)
+
+
+def multireader(files, chunks, concat_dim, **kwargs):
+    import numpy as np
+    from xarray import DataArray
+
+    dask_array = dask_imread(files, **kwargs)
+
+    ny, nx = dask_array.shape[1:3]
+    coords = {'y': np.arange(ny),
+              'x': np.arange(nx)}
+    if isinstance(concat_dim, list):
+        dims = ('dim_0', 'y', 'x')
+    else:
+        dims = (concat_dim, 'y', 'x')
+
+    if len(dask_array.shape) == 4:
+        nband = dask_array.shape[3]
+        coords['band'] = np.arange(nband)
+        dims += ('band',)
+
+    return DataArray(dask_array, coords=coords, dims=dims).chunk(chunks=chunks)
+
+
+class ImageSource(DataSourceMixin, PatternMixin):
+    """Open a xarray dataset from image files.
+
+    This creates an xarray.DataArray or an xarray.Dataset.
+    See http://scikit-image.org/docs/dev/api/skimage.io.html#skimage.io.imread
+    for the file formats supported, and
+    http://docs.dask.org/en/latest/array-api.html#dask.array.image.imread
+    for possible extra arguments.
+
+    NOTE: Although ``skimage.io.imread`` is used by default, any reader function which
+    accepts a filename and outputs a numpy array can be used instead.
```

Comment:
Since you can describe the inputs, I would explicitly say 

```
    imread: func or None
        a function such that func(file-like-object) returns a 2D or 3D array. If None, defaults to ``skimage.io.imread``
```

Reply:
yes